File: 8_kyu/CSV_representation_of_array.go
package kata

import (
	"bytes"
	"fmt"
)

// Writes integer array CVS formated in string.
func ToCsvText(array [][]int) string {
	last_row := len(array) - 1
	if last_row < 0 {
		return ""
	}
	last_col := len(array[0]) - 1
	var accu bytes.Buffer
	for row_index, row := range array {
		for col_index, nr := range row {
			fmt.Fprint(&accu, nr)
			if col_index < last_col {
				accu.WriteRune(',')
			} else if row_index < last_row {
				accu.WriteRune('\n')
			}
		}
	}
	return accu.String()
}

File: 8_kyu/CSV_representation_of_array.rs
use std::fmt::Write;

// Converts 2D array of integers into CSV formated string.
// Separates items by ',' and rows by '\n'.
fn to_csv_text(array: &[Vec<i8>]) -> String {
    if array.is_empty() || array[0].is_empty() {
        return String::new();
    }
    let last_row = array.len() - 1;
    let last_col = array[0].len() - 1;
    // estimate capacity with 3 decimals per number
    let mut accu = String::with_capacity((last_row + 1) * (last_col + 1) * 4);
    for (row, numbers) in array.iter().enumerate() {
        assert!(numbers.len() == last_col + 1);
        for (col, item) in numbers.iter().enumerate() {
            write!(accu, "{}", item).expect("out of memory");
            if col < last_col {
                accu.push(',')
            } else if row < last_row {
                accu.push('\n')
            }
        }
    }
    accu
}
